os/se/ameba/security_ameba_wrapper_tz.c

```diff
@@ -0,0 +1,763 @@
+
+#define HAL_COPY_PRIV_DATA(in, out, len)                \
+	do {                                                \
+		if (in->priv == NULL) {                         \
+			return HAL_INVALID_ARGS;                    \
+		}                                               \
+		memcpy(in->priv, out, len);                     \
+		in->priv_len = len;                             \
+	} while (0)
+
+/* Max Key or Secure Storage Index */
+#define KEY_STORAGE_INDEX_MAX 32
+#define HAL_MAX_RANDOM_SIZE 256
+
+/* Factory Key and Cert define */
+#define FACTORYKEY_RTL_DEVICE (0x00010120)
+#define FACTORYKEY_RTL_CERT (0x00010122)
+
+/* Factory Cert Key */
+#define FACTORY_CERT_ADDR CONFIG_SE_FACTORY_CERT_ADDRESS
+#define FACTORY_KEY_ADDR CONFIG_SE_FACTORY_KEY_ADDRESS
+
+/* Secure Storage Variable */
+#define SEC_STORE_OFFSET CONFIG_SE_SSTORAGE_ADDRESS
+
+/* Secure efuse key location */
+#define SAMSUNG_KEY_ADDR 0x150
+
+/* Non-Secure Data buff */
+#define NS_BUF_LEN 4096
+
+typedef struct {
+	void *ns_func_s;
+	uint8_t *buf;
+	uint32_t buf_len;
+} ns_passin_struc;
+
+typedef struct {
+	hal_data *input_data;
+	hal_data *output_data;
+} inout_struc;
+
+typedef struct {
+	uint32_t factorykey_device_idx;
+	uint32_t factorykey_cert_idx;
+	uint32_t factory_cert_addr;
+	uint32_t factory_key_addr;
+	uint32_t sstorage_addr;
+	uint32_t samsung_key_addr;
+} factory_struc;
+
+uint8_t *ns_buf;
+
+extern void* rtl_set_ns_func(void);
+
+/**
+ * Helper function
+ */
+static uint32_t rtl_index_convt(uint32_t key_idx)
+{
+	uint32_t ret_key_idx = KEY_STORAGE_INDEX_MAX;
+
+	/* Because Ameba SE storing Key in array, array start from Zero */
+	if (key_idx != 0) {	/* Index not Zero */
+		/* Valid  Index are 0 - 31, Total 32 Index slot */
+		ret_key_idx = key_idx - 1;	/* Index minus One */
+	}
+	return ret_key_idx;
+}
+
+int se_ameba_hal_init(hal_init_param *params)
+{
+	AWRAP_ENTER;
+
+	int ret;
+	factory_struc input_data;
+	ns_passin_struc ns_passin;
+
+	/* Assign Pass in function and buff */
+	ns_buf = (uint8_t *)kmm_malloc(NS_BUF_LEN);
+	if (ns_buf == NULL) {
+		ret = HAL_NOT_ENOUGH_MEMORY;
+		goto exit;
+	}
+	ns_passin.ns_func_s = rtl_set_ns_func();
+	ns_passin.buf = ns_buf;
+	ns_passin.buf_len = NS_BUF_LEN;
+
+	/* Faactory Key and Cert address (Flash) */
+	input_data.factorykey_device_idx = FACTORYKEY_RTL_DEVICE;
+	input_data.factorykey_cert_idx = FACTORYKEY_RTL_CERT;
+
+	/* Faactory Key and Cert address (Flash) */
+	input_data.factory_cert_addr = FACTORY_CERT_ADDR;
+	input_data.factory_key_addr = FACTORY_KEY_ADDR;
+
+	/* Secure Storage Base address (Flash) */
+	input_data.sstorage_addr = SEC_STORE_OFFSET;
+	input_data.samsung_key_addr = SAMSUNG_KEY_ADDR;
+
+	ret = ameba_hal_init(params, &input_data, &ns_passin);
+
+exit:
+	if (ret != HAL_SUCCESS) {
+		sedbg("RTL SE failed (%zu)\n", ret);
+	}
+	return ret;
+}
+
+int se_ameba_hal_deinit(void)
+{
+	AWRAP_ENTER;
+	int ret;
+
+	ret = ameba_hal_deinit();
+
+	if (ret == HAL_SUCCESS) {
+		if (ns_buf != NULL) {
+			kmm_free(ns_buf);
```

Comment:
Because you use this variable as global, you should set it with NULL for exception.
Let me give you one example.
1. se_ameba_hal_init to use SE
2. use SE
3. se_ameba_hal_deinit to finish usages
4. in other thread, allocate the same address with ns_buf used
5. se_ameba_hal_deinit is called again to start using (by mistake or by intended)
If ns_buff is a local variable, then setting to NULL is unnecessary. But it is a global so that it could have wrong address which is already reallocated to other thread.

Reply:
Modified.